config/abilities/provider_member.rb

```diff
@@ -26,7 +26,16 @@
     can :create, Account
     can :update, Account if account.provider_can_use?(:service_permissions)
 
-    can %i[read show edit update], Cinstance, user.accessible_cinstances.where_values_hash
+    # Using historical optimized permission and leave canonical way commented out below
+    # The resulting hash presently is something like {"type"=>"Cinstance", "service_id"=>[ids..]}
+    # Seems like in general we are moving towards accessing service through plan.
+    # Also this doesn't check account of the service but `member_permission_service_ids` will not
+    # return foreign IDs. Also tenant_id checking should not allow foreign objects access.
+    can %i[read show edit update], Cinstance, Cinstance.permitted_for(user).where_values_hash
+    # can %i[read show edit update], Cinstance, user.accessible_cinstances do |cinstance|
+    #   cinstance.plan&.issuer_type == "Service" && cinstance.plan.issuer.account == user.account &&
+    #     (!user.forbidden_some_services? || user.member_permission_service_ids.include?(cinstance.plan.issuer.id))
+    # end
```

Comment:
What's the difference between the new one, the old one and the commented one?

Reply:
the commented out uses a block which is not recommended. I believe more description is in the commit messages. But for the time being, I'm looking at other things.

Reply:
fix user can access Cinstance
    
    The `#where_values_hash` method does not support joins and sub-queries.
    
    Originally the `account.provided_cinstances` part was ignored because it
    was JOINs. With the update to sub-queries, it turned into `plan_id: nil`
    which is incorrect and broke the logic.
    
    So now we keep logic as previously by resorting only to the
    `Cinstance.permitted_for` part of the query.
    
    This relies on the fact that `Cinstance.plan.issuer` is set as
    `Cinstance.service` when that issuer is a service.
    
    Also relies on the fact that `User.member_permission_service_ids` will
    not set to ids of services that don't belong to the account.
    
    Which may not be ideal but allows for permission checking without extra
    database queries.